### ISSUE TITLE
Idea: Remove the mutation GraphQL specification requirements?

### ISSUE THREAD

**Issue by ianstormtaylor** [facebook/relay#825]
One of the things I found annoying while trying to implement mutations is how much opinion Relay inflicts on my GraphQL schema—having to create `Input` and `Payload` definitions for each operation is really verbose, very Relay-specific, and not necessarily that “clean” of an API.

For example, ignoring Relay, it might be nice for a mutation in the GraphQL schema to look like:

```
type Mutation {
  createTeam(userId: ID!, team: JSON!): Team
}
```

That would be a fairly simple `createTeam` mutation, that is easy to understand. You pass the `userId` of the user you’re creating the team for, and a `team` JSON object containing team attributes (name, billing email, etc.). It maps fairly simply to a `POST /users/:userId/teams` call of traditional REST APIs (not that that’s a requirement, but REST does have a lot of good patterns).

Unfortunately, implementing that same mutation for Relay requires the GraphQL schema to be very Relay-specifc instead:

```
type Mutation {
  createTeam(input: CreateTeamInput!): CreateTeamPayload
}

input CreateTeamInput {
  userId: ID!
  team: JSON!
  clientMutationId: String!
}

type CreateTeamPayload {
  team: Team
  clientMutationId: String!
}
```

A lot more verbose, but also less easily used by other, non-Relay, GraphQL implementations now. Or at least without them having to adopt Relay’s naming opinions.

It seems like one reason the `Input` is used is so that creating the `$input` variable is as easy as possible for the client code. But it doesn’t seem like this should come at the expense of changing the schema itself. It seems like it would still be easy to implement arbitrary variable names, and have them be stringified properly by the query configuration logic in Relay. Such that passing something like:

``` js
{
  userId: 'user_1',
  team: {
    name: 'Team',
    email: 'team@example.com'
  }
}
```

Could be stringified into a mutation of:

```
mutation CreateTeam($userId: String, $team: JSON) {
  createTeam(userId: $userId, team: $team) ...
}
```

Another reason for the `Input` and `Payload` is to attach a `clientMutationId`, which the documentation suggests is  `used to reconcile requests and responses`. But this seems like something that is NetworkLayer specific, and could be better implemented as a separate parameter to any HTTP calls made to the server, and even something that might need to be use-case specific. 

For example, I already support a `?requestId` parameter on my server for that exact purpose, and others might have differently named solutions. If all that’s needed is a unique identifier for requests, couldn’t it be implemented decoupled from the GraphQL part of the request? Potentially even implemented as an HTTP header instead of as part of the parameters of a request. Seems like something that’s going to be very server-specific, and so should be an option for the NetworkLayer, but have nothing to do with the GraphQL schema itself?

**Comment by ianstormtaylor** [comment]
Realized I forgot to add another fake-looking reason for the `clientMutationId`, to separate individual mutations from each other. But since they'd have to be named if multiple mutations were passed, this shouldn't actually be a need. The `clientMutationId` seems to really be a `clientRequestId`, which is why it would be even more awesome to remove it from the GraphQL schema part of the request.

Adding more mutations today and really feeling the extra complexity of having to define individual `Input` and `Payload` types for each and every mutation in the API. I'm thinking my ideal would be something like:

```
type Payload {
  user: User
  team: Team
  item: Item
  ...
}

type Mutation {
  createUser(attributes: JSON): Payload
  updateUser(userId: ID, attributes: JSON): Payload
  deleteUser(userId: ID): Payload

  createTeam(userId: ID, attributes: JSON): Payload
  updateTeam(teamId: ID, attributes: JSON): Payload
  deleteTeam(teamId: ID): Payload

  createItem(teamId: ID, attributes: JSON): Payload
  updateItem(itemId: ID, attributes: JSON): Payload
  deleteTeam(itemId: ID): Payload

  ...
}
```

Here with `Payload` acting very similarly to `Query`, as a generic way to return multiple types of objects across all mutation requests. Such that for `createTeam` it could return `{ user, team }` so that the client can do whatever it pleases. (I'd use `Query` itself actually, but it seems a bit off since they shouldn't be able to dig into unrelated objects, or at least it would seem to encourage that.)

Note: `JSON` and not using `!` for required attributes are just because in this case I'm sharing the same database module across a GraphQL and REST API, and would prefer to keep all of the validation in that module itself, so that error handling is consistent across both APIs. But if I wasn't, it would totally make sense take things like `createTeam(userId: ID, teamName: String, teamEmail: String): Payload` or similar.

Just really wish the extra wrapper types were gone is all! Since right now that above code turns into something like:

```
input CreateUserInput {
  attributes: JSON
  clientMutationId: String!
}

input UpdateUserInput {
  userId: ID
  attributes: JSON
  clientMutationId: String!
}

input DeleteUserInput {
  userId: ID
  clientMutationId: String!
}

type CreateUserPayload {
  user: User
  clientMutationId: String!
}

type UpdateUserPayload {
  user: User
  clientMutationId: String!
}

type DeleteUserPayload {
  user: User
  clientMutationId: String!
}

input CreateTeamInput {
  userId: ID
  attributes: JSON
  clientMutationId: String!
}

input UpdateTeamInput {
  teamId: ID
  attributes: JSON
  clientMutationId: String!
}

input DeleteTeamInput {
  teamId: ID
  clientMutationId: String!
}

type CreateTeamPayload {
  user: User
  team: Team
  clientMutationId: String!
}

type UpdateTeamPayload {
  user: User
  team: Team
  clientMutationId: String!
}

type DeleteItemPayload {
  team: Team
  item: Item
  clientMutationId: String!
}

input CreateItemInput {
  teamId: ID
  attributes: JSON
  clientMutationId: String!
}

input UpdateItemInput {
  itemId: ID
  attributes: JSON
  clientMutationId: String!
}

input DeleteItemInput {
  itemId: ID
  clientMutationId: String!
}

type CreateItemPayload {
  team: Team
  item: Item
  clientMutationId: String!
}

type UpdateItemPayload {
  team: Team
  item: Item
  clientMutationId: String!
}

type DeleteItemPayload {
  team: Team
  item: Item
  clientMutationId: String!
}

...

type Mutation {
  createUser(input: CreateUserInput!): CreateUserPayload
  updateUser(input: UpdateUserInput!): UpdateUserPayload
  deleteUser(input: DeleteUserInput!): DeleteUserPayload

  createTeam(input: CreateTeamInput!): CreateTeamPayload
  updateTeam(input: UpdateTeamInput!): UpdateTeamPayload
  deleteTeam(input: DeleteTeamInput!): DeleteTeamPayload

  createItem(input: CreateItemInput!): CreateItemPayload
  updateItem(input: UpdateItemInput!): UpdateItemPayload
  deleteItem(input: DeleteItemInput!): DeleteItemPayload

  ...
}
```


Let me know if I'm missing something there though, happy to think through more.

**Comment by wincent** [comment]
Thanks for this, and sorry for taking so long to explicitly respond to it.

There are really two issues here:

- One is about client mutation IDs, and that is addressed in #1083. There is some context in that thread, but the short version is that it's in there because of legacy reasons, where our iOS clients at FB needed it for book-keeping purposes. As noted in there, however, we'll can lift the restriction in the new Relay core/APIs.
- The other is about the ergonomics of creating mutations. This is another thing that will be largely addressed in the new core: while we will continue to support the "Relay spec" version of mutations out of the box, we will also supply lower-level tools and building blocks that allow people to construct their mutations pretty much any way they want. The new core and APIs aren't fully released yet, but if you watch the `master` branch you can see the pieces that we've started to roll out, and will continue doing over the coming weeks/months.

I'm going to suggest that we close this in the interests of spring-cleaning. If you still have concerns once the new APIs are out, we should start a new issue for that. Thanks for posting in such a detailed and thoughtful way!
